### PR TITLE
fix(cssClasses): Fixed duplication of classNames

### LINKED ISSUE #388: Scope CSS classes
In our widgets, we're not scoping the `cssClasses` declaration, which is reusing the parameter passed to the widget.
The result is that if render is called multiple times, we end up with duplicated class names.

### PR DESCRIPTION
**Pull request by Jerskouille**
Closes #388

### CODE REVIEW

**Comment by vvo** [comment]
Even if the tests are green, we need new tests to ensure we are fixing the bug

**Comment by vvo** [comment]
So it means having tests calling render multiple times on same element

**Comment by Jerskouille** [comment]
For each widget? Do we really need that?

**Comment by vvo** [comment]
Well otherwise I can revert your commit and test still green. Every bugfix=test

**Comment by Jerskouille** [comment]
We didn't test anything when we introduced `dedupe` to solve exactly this issue.

File: decorators/headerFooter.js [hunk 1]
```diff
@@ -14,15 +14,15 @@ function headerFooter(ComposedComponent) {
       if (!templates || !templates[type]) {
         return null;
       }
-      let className = cx(this.props.cssClasses[type], `ais-${type}`);
+      let cssClass = cx(this.props.cssClasses[type], `ais-${type}`);
```
Comment (pixelastic) [inline decorators/headerFooter.js]: Why renaming to `cssClass`? I try to keep the name of the variables equals to the name of the props they will be bound to, to make it easier to understand

**Comment by pixelastic** [comment]
My bad, I was the one introducing `dedupe` and I did not add any tests to it.

I agree we need to have tests to prove that we actually fixed a bug and prevent future regressions. But adding the same tests to all widgets is cumbersome. Does it mean we'll have to add this boilerplate test to all new widgets?

I have no answer to that unfortunatly.

File: decorators/headerFooter.js (hunk 1 above)
Comment (Jerskouille) [inline decorators/headerFooter.js]: Consistency with `cssClasses` used in the same file, but I can remove it.

Comment (Jerskouille) [inline decorators/headerFooter.js]: We're using `className={cssClasses.root}`, made more sense to me to have `className={cssClass}` to match these.

**Comment by Jerskouille** [comment]
If you're fine with #397 , could we merge it before this one ?

**Comment by vvo** [comment]
> If you're fine with #397 , could we merge it before this one ?

Yes.

What about testing here? Do we do it or in another PR?

**Comment by Jerskouille** [comment]
I've rebased on #397 to do the testing since it modified the JSX, but the tests are ready.

File: widgets/hits-per-page-selector/__tests__/hits-per-page-selector-test.js [hunk 2]
```diff
@@ -73,9 +74,11 @@ describe('hitsPerPageSelector()', () => {
       ],
       setValue: () => {}
     };
-    expect(ReactDOM.render.calledOnce).toBe(true, 'ReactDOM.render called once');
+    expect(ReactDOM.render.calledTwice).toBe(true, 'ReactDOM.render called twice');
     expect(ReactDOM.render.firstCall.args[0]).toEqualJSX(<Selector {...props} />);
     expect(ReactDOM.render.firstCall.args[1]).toEqual(container);
+    expect(ReactDOM.render.secondCall.args[0]).toEqualJSX(<Selector {...props} />);
```
Comment (vvo) [inline widgets/hits-per-page-selector/__tests__/hits-per-page-selector-test.js]: Good, in the end it was not so much boilerplate code to add to test this bug. In the future we might just have a `BaseWidget` to break a little bit this redundant logic we have (cssClasses)

File: decorators/headerFooter.js (hunk 1 above)
Comment (vvo) [inline decorators/headerFooter.js]: I would go for `className` also.

**Comment by Jerskouille** [comment]
Well, yes and no, because here we were just missing a `var` keyword, which would have been missing even if we embedded all the logic in a common function, the issue was present in all widgets.

There's a choice to make, and a little bit of consistency to have here. Do you really want to unit test everything ? If yes, we can't test only one for code that is copy/pasted everywhere. But then we shouldn't only add tests on bug-fixes, but also on every `feat` and `chore` PR.

By the way, I didn't add tests for widgets that I fixed but didn't have a test file.

**Comment by vvo** [comment]
> But then we shouldn't only add tests on bug-fixes, but also on every feat and chore PR

Oh yes we should have tests for feats that for sure. Then for chore() it depends on the chore.

**Comment by vvo** [comment]
> which would have been missing even if we embedded all the logic in a common function

Yes and we would have had to only add a test to this common function. In the underlying widget you just assume the function is called.

This is unit testing yes so it does not ensures everything, it's hints and ease of developments for us.

File: widgets/hierarchical-menu/hierarchical-menu.js [hunk 3]
```diff
@@ -81,7 +81,7 @@ function hierarchicalMenu({
         templates
       });
 
-      cssClasses = {
```
Comment (vvo) [inline widgets/hierarchical-menu/hierarchical-menu.js]: Maybe we could do this:

```js
function hierarchicalMenu({
    container,
    attributes = [],
    separator,
    limit = 100,
    sortBy = ['name:asc'],
    cssClasses: userCssClasses = {},
    hideContainerWhenNoResults = true,
    templates = defaultTemplates,
    transformData
  }) {
    // ....
    // var cssClasses = {// use userCssClasses safely}
};
```

**Comment by Jerskouille** [comment]
> Yes and we would have had to only add a test to this common function. In the underlying widget you just assume the function is called.

vs

> Well otherwise I can revert your commit and test still green. Every bugfix=test

Hum ?

File: widgets/hierarchical-menu/hierarchical-menu.js (hunk 3 above)
Comment (Jerskouille) [inline widgets/hierarchical-menu/hierarchical-menu.js]: I'm not sure about this syntax, but here, aren't both `userCssClasses` and `cssClasses` defined when entering the function?

Comment (vvo) [inline widgets/hierarchical-menu/hierarchical-menu.js]: I was wondering also, I tested it that's fine: https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=function%20hierarchicalMenu(%7B%0A%20%20%20%20container%2C%0A%20%20%20%20attributes%20%3D%20%5B%5D%2C%0A%20%20%20%20separator%2C%0A%20%20%20%20limit%20%3D%20100%2C%0A%20%20%20%20sortBy%20%3D%20%5B'name%3Aasc'%5D%2C%0A%20%20%20%20cssClasses%3A%20userCssClasses%20%3D%20%7B%7D%2C%0A%20%20%20%20hideContainerWhenNoResults%20%3D%20true%2C%0A%20%20%20%20templates%20%3D%20defaultTemplates%2C%0A%20%20%20%20transformData%0A%20%20%7D)%20%7B%0A%20%20%20%20%0A%7D%3B

Comment (Jerskouille) [inline widgets/hierarchical-menu/hierarchical-menu.js]: Alright, why not then, I'll do that

**Comment by vvo** [comment]
@Jerskouille needs rebase since previous PR about price range was merged, sorry abt that. It happens

**Comment by vvo** [comment]
Thanks a lot for the bugfix @Jerskouille, we can have long unit test talks discussions soon :D

**Comment by redox** [comment]
Great job @Jerskouille :+1: :100:

**Pull request merged**